src/_InputField/InputField.tsx

```diff
@@ -0,0 +1,116 @@
+import {uniqueId} from '../utils/uniqueId'
+import InputFieldCaption from './InputFieldCaption'
+import InputFieldLabel from './InputFieldLabel'
+import InputFieldValidation from './InputFieldValidation'
+import {Slots} from './slots'
+import ValidationAnimationContainer from './ValidationAnimationContainer'
+export interface Props<T = Record<string, FormValidationStatus>> {
+  children?: React.ReactNode
+  /**
+   * Whether the field is ready for user input
+   */
+  disabled?: boolean
+  /**
+   * The unique identifier for this field. Used to associate the label, validation text, and caption text
+   */
+  id?: string
+  /**
+   * Whether this field must have a value for the user to complete their task
+   */
+  required?: boolean
+  /**
+   * A map of validation statuses and their associated validation keys. When one of the validation keys is passed to the `validationResult` prop,
+   * the associated validation message will be rendered in the correct style
+   */
+  validationMap?: T
+  /**
+   * The key of the validation message to show
+   */
+  validationResult?: keyof T
+}
+
+type InputFieldValidationProps = ComponentProps<typeof InputFieldValidation>
+export interface InputFieldContext
+  extends Pick<Props<Record<string, FormValidationStatus>>, 'disabled' | 'id' | 'required'> {
+  captionId: string
+  validationMessageId: string
+}
+
+const InputField = <T extends Record<string, FormValidationStatus>>({
+  children,
+  disabled,
+  id: idProp,
+  required,
+  validationMap,
+  validationResult
+}: Props<T>) => {
+  const id = idProp || uniqueId()
+  const validationChildren: React.ReactElement<InputFieldValidationProps>[] | undefined | null = React.Children.map(
+    children,
+    child =>
+      React.isValidElement<InputFieldValidationProps>(child) && child.type === InputFieldValidation ? child : null
+  )?.filter(Boolean)
+  const captionChildren: React.ReactElement[] | undefined | null = React.Children.map(children, child =>
+    React.isValidElement(child) && child.type === InputFieldCaption ? child : null
+  )?.filter(Boolean)
+  const validationChildToRender = validationChildren?.find(child => child.props.validationKey === validationResult)
+  const validationMessageId = validationChildToRender ? `${id}-validationMsg` : undefined
+  const captionId = captionChildren?.length ? `${id}-caption` : undefined
+
+  return (
+    <Slots
+      context={{
+        captionId,
+        disabled,
+        id,
+        required,
+        validationMessageId
+      }}
+    >
+      {slots => {
+        const isLabelHidden = React.isValidElement(slots.Label) && slots.Label.props.visuallyHidden
+        if (slots.Input && !slots.Label) {
+          // eslint-disable-next-line no-console
+          console.error(
+            `The input field with the id ${id} MUST have a Label child (e.g.: <TextInputField.Label>).\n\nIf you want to hide the label, pass the 'visuallyHidden' prop to the Label component.`
+          )
```

Comment:
Would you like to throw an Error instead? 

```
throw ... raises an exception in the current code block and causes it to exit, or to flow to next catch statement if raised in a try block.

console.error just prints out a red message to the browser developer tools javascript console and does not cause any changes of the execution flow.
```
source: https://stackoverflow.com/a/25377160

I throw an Error in ActionList v2 here: https://github.com/primer/react/blob/main/src/ActionList2/Selection.tsx#L20

Reply:
I was thinking that throwing an error might be too aggressive incase an unlabeled input makes it into production somehow.

Maybe we can safely assume that our CI will catch it though

Reply:
Fair enough :)